src/bundle/streaming_reader.zig

```diff
@@ -84,172 +79,84 @@ pub const DecompressingHashReader = struct {
     pub fn deinit(self: *Self) void {
         _ = c.ZSTD_freeDCtx(self.dctx);
         self.allocator_ptr.free(self.in_buffer);
-        self.allocator_ptr.free(self.out_buffer);
+        self.allocator_ptr.free(self.interface.buffer);
     }
 
     fn stream(r: *std.Io.Reader, w: *std.Io.Writer, limit: std.Io.Limit) std.Io.Reader.StreamError!usize {
-        const self: *Self = @alignCast(@fieldParentPtr("interface", r));
-        const dest = limit.slice(try w.writableSliceGreedy(1));
-        const n = self.read(dest) catch |err| switch (err) {
-            error.DecompressionFailed, error.HashMismatch => return std.Io.Reader.StreamError.ReadFailed,
-            error.UnexpectedEndOfStream => return std.Io.Reader.StreamError.EndOfStream,
-            error.OutOfMemory => return std.Io.Reader.StreamError.ReadFailed,
-        };
-        if (n == 0) {
-            return std.Io.Reader.StreamError.EndOfStream;
+        // This implementation just adds the decompressed data to the buffer and returns 0.
+        // This simplifies the logic a bit which is encouraged by the Zig reader API.
+        _ = w;
+        _ = limit;
+        if (r.end == r.seek) {
+            r.end = 0;
+            r.seek = 0;
         }
-        w.advance(n);
-        return n;
-    }
-
-    fn discard(r: *std.Io.Reader, limit: std.Io.Limit) std.Io.Reader.Error!usize {
         const self: *Self = @alignCast(@fieldParentPtr("interface", r));
 
-        var total: usize = 0;
-        var remaining: ?usize = limit.toInt();
+        var in_writer = std.Io.Writer.fixed(self.in_buffer[self.in_end..]);
+        var reached_end = false;
+        const bytes_read = self.input_reader.stream(&in_writer, std.Io.Limit.limited(self.in_buffer.len)) catch |err| switch (err) {
+            error.EndOfStream => blk: {
+                reached_end = true;
+                break :blk 0;
+            },
+            error.ReadFailed => return error.ReadFailed,
+            error.WriteFailed => unreachable, // fixed buffer writer doesn't fail
+        };
 
-        // Consume any buffered output data first.
-        if (self.out_pos < self.out_end) {
-            const available = self.out_end - self.out_pos;
-            const to_consume = if (remaining) |rem| @min(available, rem) else available;
-            self.out_pos += to_consume;
-            total += to_consume;
-            if (remaining) |*rem| {
-                rem.* -= to_consume;
-                if (rem.* == 0) return total;
+        if (reached_end) {
+            // verify hash if not already done
+            if (!self.hash_verified) {
+                var actual_hash: [32]u8 = undefined;
+                self.hasher.final(&actual_hash);
+                if (std.mem.eql(u8, &actual_hash, &self.expected_hash)) {
+                    self.hash_verified = true;
```

Comment:
I'm kinda surprised this is just a bool on the struct and not an error state if unverified at the end.